### ISSUE TITLE
Suggestion - Automatically Check out folders/files from TFS when refreshing objects

### ISSUE THREAD

**Issue by CorsairRO** [ErikEJ/EFCorePowerTools#1230]
Hello again,

Just a suggestion there ( as i noticed i had to manually checkout all the files before refreshing my model )
Including efpt.config.json.

If it is easy to do, it would be a nice addition to the product.
Otherwise, no problem, please just close the ticket, i have placed all Generated files in a folder, so i can do that manually before refreshing models

### Steps to reproduce

Add a model and generated models to TFS ( i use Azure DevOps with TFS )
Use refresh/edit  contextmenu on the .json file.
Complains files are readonly.

I did this before with edmx customized .tt files, where i have used something in VS, like IHost or similar.
But i think it comes in old good plain .tt where is 1 atribute to set hostspecific="true"
So if MS decided scaffolding goes away from .tt then this is it

thank you
have a nice day, and a Merry Christmas and a Safe New Year

### Further technical details

EF Core version in use: (e.g. EF Core 5)

EF Core Power Tools version: (found in About dialog - blue questionmark icon on context menu)

Database engine: (SQL Server, Azure SQL, Postgres, Oracle, Firebird, SQLite, MySQL)

Visual Studio version: (e.g. Visual Studio 2019 16.11)

**Comment by ErikEJ** [comment]
Are you using TFVC with Azure DevOps? You should convert to git 😄 

Curently, with the interop library I am using, there is no feature to do this.

**Comment by CorsairRO** [comment]
hello, yes, indeed i am using TVFC as i cannot understand how i can use git in a my real environment.

i have a team where each works on separate parts, so i wanted to use git but it DOENST have exclusive checkouts, and therefore i moved to TVFC.

As i said, no problem, it as just a suggestion, so please close it if hard to do :)

**Comment by ErikEJ** [comment]
You should just use feature branches with git!

**Comment by ErikEJ** [comment]
Looks like it could be possible: https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/issues/258#issuecomment-1011383459

**Comment by CorsairRO** [comment]
hello,

no, it is not possible at least for the moment, after investigating a bit your tool there.

Before Community toolkit answer about that i have tried in my copy of your tool to use EnvDTE from SDK project, and it was working fine, and it has DTE.SourceControl.Checkout and stuff, but i have hit the limitation from above.

Problem: Files are saved to disk by ReverseEngineerRunner that is inside RevEng.Core project that has no knowledge about any Community.Toolkit reference, and most probably it should stay like this, as it is .NET Standard, while the Extension itself is .NET Framework 4.7.2

Most notably RevEng.Core has no knowledge about Services used inside ExtensionPackage Project, and only that one uses VS Community toolkit.

Thank you for trying to help there

**Comment by CorsairRO** [comment]
Hello again,

As i am messing a lot with your tool like i said in a copy i can try to move ALL the file creations somehow up in the hierarchy of projects there.
This will also help in a better architecture let's say, as now files are written in many places in the RevEng.Core project.
Like some files in ReverseEngineerScaffolder, some in ProcedureScaffolder and so on ...
What i will try is instead of writing files directly to disk there to try to add them to a dictionary of some sort, with filename the key, and return back to the VSIX Extension that one.

Then the extension will take care of writing files to disk ALL in a single place, like that i can do something that i also love about edmx T4, using TryCheckout and File.Write ONLY if the file has changed and so on ...

I will come back to you if i end up anywhere there, and if you want you can include it in the tool

Offtopic 1: Anyway, one definitive change that i made and now it is working in my copy is that i am TOTALLY removing the sp_describe_first_resultset crap SP as it is NOT working at all for my case. Using FMTONLY ON and take First resultset is working 100%, as opposite to 10% of that proc. I dont want to add Dapper only to get the results from SP, and i dont need, as i dont have MultipleResultsets. So if i may have a suggestion this will be defintively a good addition to the tool ( maybe as a separate option or so like MARS ).

Offtopic 2: I guess you are aware that the guys from EF Core are willing to put back the T4 templating in the EF7 Plan.
So they seem to realize how useful that feature was and how simpler it makes life for people wanting to develop real time applications with more than 5 tables inside it.

Thank you for your HUGE work there to keep such a nice tool,
Will come back to you if i manage to do something,
Have a nice day

**Comment by ErikEJ** [comment]
>> Using FMTONLY ON and take First resultset is working 100%,

Yes, I am considering making that the default, and add an option to enable the old behavior (so in that case Dapper is not required)

>> them to a dictionary of some sort,

Extending the current ReverseEngineerResult to return filename and content instead of just filenames sounds like a great idea, a PR for that would be welcome.

>> EnvDTE 

I cannot use EnvDTE in an extension that supports both VS 2019 and VS 2022, but it looks like there is a solution coming for the Community Toolkit for source control.

**Comment by CorsairRO** [comment]
OK,

FMTONLY => thank you very very much

EnvDTE => I know that this one is bitchy, the Community Toolkit question is my opened one, so i will test it as i promised to the author guy from there and let him know how it works.

Extending Result PR will be welcome => Am is supposed to do something there, or you create one and i can look and mess there? Sorry for my thumbness there, just asking as never worked like i said in 15+ YY with other source control then TFS so i am not very familiar with that. But of course i would like to learn a bit of that.

thank you,
have a nice evening there

**Comment by ErikEJ** [comment]
I think a better approach would be to have the class in a nuget package.
